File: test/node/widget/account_unbinding_test.ts
import Dana from 'dana-node';
import { v4 as uuidv4 } from 'uuid';
import * as path from 'path';
import * as dotenv from 'dotenv';
import { fail } from 'assert';
import { getRequest } from '../helper/util';
import { assertResponse, assertFailResponse } from '../helper/assertion';
import { AccountUnbindingRequest, ApplyTokenRequest } from 'dana-node/widget/v1';
import { executeManualApiRequest } from '../helper/apiHelpers';
import { generateAuthCode } from '../helper/auth-utils';

dotenv.config();

const titleCase = 'AccountUnbinding';
const jsonPathFile = path.resolve(__dirname, '../../../resource/request/components/Widget.json');

const dana = new Dana({
    partnerId: process.env.X_PARTNER_ID || '',
    privateKey: process.env.PRIVATE_KEY || '',
    origin: process.env.ORIGIN || '',
    env: process.env.ENV || 'sandbox',
});

async function generateApplyToken(phoneNumber?: string, pinCode?: string): Promise<string> {
    const caseName = 'ApplyTokenSuccess';
    const requestData: ApplyTokenRequest = getRequest<ApplyTokenRequest>(jsonPathFile, "ApplyToken", caseName);

    requestData.authCode = await generateAuthCode(phoneNumber, pinCode);

    const response = await dana.widgetApi.applyToken(requestData);
    return response.accessToken;
}

describe('Account Unbinding Tests', () => {
  // Add beforeAll if any shared setup is needed in the future

  test('should successfully unbind account', async () => {
    const caseName = 'AccountUnbindSuccess';
    const requestData: AccountUnbindingRequest = getRequest<AccountUnbindingRequest>(jsonPathFile, titleCase, caseName);
    requestData.additionalInfo.accessToken = await generateApplyToken();
    requestData.additionalInfo.deviceId = "deviceid123";
    requestData.merchantId = process.env.MERCHANT_ID || '';

    try {
      const response = await dana.widgetApi.accountUnbinding(requestData);
      console.log("Response Data for Account Unbinding:", JSON.stringify(response, null, 2));
      await assertResponse(jsonPathFile, titleCase, caseName, response);
    } catch (e: any) {
      console.error("Error during Account Unbinding:", e);
      fail('Account unbinding test failed: ' + (e.message || e));
    }
  });

  test.skip('should fail to unbind account when access token does not exist', async () => {
    const caseName = 'AccountUnbindFailAccessTokenNotExist';
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);
    try {
      const response = await dana.widgetApi.accountUnbinding(requestData);
      await assertFailResponse(jsonPathFile, titleCase, caseName, response);
      fail('Expected an error but the API call succeeded');
    } catch (e: any) {
      // Optionally assert error structure here if needed
    }
  });

  test.skip('should fail to unbind account with invalid user status', async () => {
    const caseName = 'AccountUnbindFailInvalidUserStatus';
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);
    try {
      const response = await dana.widgetApi.accountUnbinding(requestData);
      await assertFailResponse(jsonPathFile, titleCase, caseName, response);
      fail('Expected an error but the API call succeeded');
    } catch (e: any) {
      // Optionally assert error structure here if needed
    }
  });

  test.skip('should fail to unbind account with invalid params', async () => {
    const caseName = 'AccountUnbindFailInvalidParams';
    const requestData: any = getRequest(jsonPathFile, titleCase, caseName);
    try {
      const response = await dana.widgetApi.accountUnbinding(requestData);
      await assertFailResponse(jsonPathFile, titleCase, caseName, response);
      fail('Expected an error but the API call succeeded');
    } catch (e: any) {
      // Optionally assert error structure here if needed
    }
  });
});

File: test/go/widget/account_unbinding_test.go
package widget_test

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"uat-script/helper"
	widget_helper "uat-script/widget"

	"github.com/dana-id/dana-go/v2/widget/v1"
)

const (
	widgetAccountUnbindingCase = "AccountUnbinding"
)

// AccountUnbinding
func TestAccountUnbindSuccess(t *testing.T) {
	caseName := "AccountUnbindSuccess"

	// Get auth code via OAuth flow
	redirectUrlAuthCode, _ := widget_helper.GetRedirectOauthUrl(
		helper.TestConfig.PhoneNumber,
		helper.TestConfig.PIN,
	)
	authCode, _ := widget_helper.GetAuthCode(
		helper.TestConfig.PhoneNumber,
		helper.TestConfig.PIN,
		redirectUrlAuthCode)

	// Exchange auth code for a real access token via ApplyToken API (matching PHP setUpBeforeClass)
	ctx := context.Background()
	applyTokenReq := widget.NewApplyTokenAuthorizationCodeRequest("AUTHORIZATION_CODE", authCode)
	applyTokenReqValue := widget.ApplyTokenAuthorizationCodeRequestAsApplyTokenRequest(applyTokenReq)
	applyTokenResp, _, err := helper.ApiClient.WidgetAPI.ApplyToken(ctx).ApplyTokenRequest(applyTokenReqValue).Execute()
	if err != nil {
		t.Fatalf("Failed to obtain access token: %v", err)
	}
	accessToken := applyTokenResp.GetAccessToken()
	fmt.Printf("Obtained access token: %s\n", accessToken)

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetAccountUnbindingCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	jsonDict["additionalInfo"] = map[string]interface{}{
		"accessToken": accessToken,
		"deviceId":    helper.TestConfig.DeviceID,
	}
	jsonDict["merchantId"] = helper.TestConfig.MerchantID

	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	accountUnbindingRequest := widget.AccountUnbindingRequest{}
	err = json.Unmarshal(jsonBytes, &accountUnbindingRequest)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Make the API call using the Widget SDK
	apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.AccountUnbinding(ctx).AccountUnbindingRequest(accountUnbindingRequest).Execute()
	if err != nil {
		t.Fatalf("API request failed: %v", err)
	}
	defer httpResponse.Body.Close()

	responseJSON, err := apiResponse.MarshalJSON()
	if err != nil {
		t.Fatalf("Failed to convert response to JSON: %v", err)
	}

	err = helper.AssertResponse(
		helper.TestConfig.JsonWidgetPath,
		widgetAccountUnbindingCase,
		caseName,
		string(responseJSON),
		nil,
	)
	if err != nil {
		t.Fatal(err)
	}
}

func TestAccountUnbindFailInvalidUserStatus(t *testing.T) {
	t.Skip("Skipping test AccountUnbindFailInvalidUserStatus")
	caseName := "AccountUnbindFailInvalidUserStatus"
	redirectUrlAuthCode, _ := widget_helper.GetRedirectOauthUrl(
		helper.TestConfig.PhoneNumber,
		helper.TestConfig.PIN,
	)
	authCode, _ := widget_helper.GetAuthCode(
		helper.TestConfig.PhoneNumber,
		helper.TestConfig.PIN,
		redirectUrlAuthCode)

	// Get the request data from JSON
	jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetAccountUnbindingCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}
	jsonDict["authCode"] = authCode

	// Marshal to JSON and unmarshal to widget SDK struct for type safety
	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	accountUnbindingRequest := &widget.AccountUnbindingRequest{}
	err = json.Unmarshal(jsonBytes, accountUnbindingRequest)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	// Execute the SDK API call with success expectation
	ctx := context.Background()

	// Make the API call using the Widget SDK
	apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.AccountUnbinding(ctx).AccountUnbindingRequest(*accountUnbindingRequest).Execute()
	if err != nil {
		t.Fatalf("API request failed: %v", err)
	}
	defer httpResponse.Body.Close()

	// Convert the response to JSON for assertion
	responseJSON, err := apiResponse.MarshalJSON()
	if err != nil {
		t.Fatalf("Failed to convert response to JSON: %v", err)
	}

	err = helper.AssertResponse(
		helper.TestConfig.JsonWidgetPath,
		widgetAccountUnbindingCase,
		caseName,
		string(responseJSON),
		nil,
	)
	if err != nil {
		t.Fatal(err)
	}
}
